### PR TITLE
PP-12173 upgrade to Dropwizard 3

### PR DESCRIPTION
**Pull request by SandorArpa**
## WHAT YOU DID

- upgrade to Dropwizard 3

### CODE REVIEW

File: .github/dependabot.yml [hunk 1]
```diff
@@ -8,9 +8,9 @@
       time: "03:00"
     ignore:
       - dependency-name: "io.dropwizard:dropwizard-dependencies"
-        # We are still preparing to upgrade to Dropwizard 3
+        # We are still preparing to upgrade to Dropwizard 4
         versions:
-          - ">= 3"
+          - ">= 4"
```
Comment (alexbishop1) [inline .github/dependabot.yml]: We can use the same comment as https://github.com/alphagov/pay-adminusers/blob/659a230d6a4d6b66cb54ec807b70caa3fc07e419/.github/dependabot.yml#L10-L13 for consistency.

File: pom.xml [hunk 2]
```diff
@@ -34,7 +20,149 @@
         <PACT_CONSUMER_VERSION/>
         <PACT_CONSUMER_TAG/>
     </properties>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>io.dropwizard</groupId>
+                <artifactId>dropwizard-dependencies</artifactId>
+                <version>3.0.6</version>
```
Comment (alexbishop1) [inline pom.xml]: There’s a 3.0.7 now.

File: pom.xml [hunk 3]
```diff
@@ -34,7 +20,149 @@
...
+            <artifactId>commons-lang3</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.amazonaws</groupId>
+            <artifactId>aws-java-sdk-sqs</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.google.inject.extensions</groupId>
+            <artifactId>guice-assistedinject</artifactId>
+        </dependency>
```
Comment (alexbishop1) [inline pom.xml]: Can you put this next to the other Guice dependencies?

File: pom.xml [hunk 4]
```diff
@@ -398,37 +397,15 @@
...
-                    <artifactId>jetty-servlets</artifactId>
-                </exclusion>
-                <exclusion>
-                    <groupId>org.eclipse.jetty</groupId>
-                    <artifactId>jetty-webapp</artifactId>
-                </exclusion>
-            </exclusions>
+            <groupId>org.wiremock</groupId>
+            <artifactId>wiremock-standalone</artifactId>
+            <version>3.4.1</version>
```
Comment (alexbishop1) [inline pom.xml]: We’re up to 3.5.2 now.

File: pom.xml [hunk 5]
```diff
@@ -336,41 +311,65 @@
...
+            <groupId>org.testcontainers</groupId>
+            <artifactId>postgresql</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.dropwizard</groupId>
+            <artifactId>dropwizard-testing</artifactId>
+            <scope>test</scope>
+        </dependency>
+
```
Comment (alexbishop1) [inline pom.xml:383]: Rogue blank line?

File: pom.xml [hunk 6]
```diff
@@ -398,37 +397,15 @@
...
+            <version>3.4.1</version>
             <scope>test</scope>
         </dependency>
         <dependency>
-            <groupId>io.dropwizard</groupId>
-            <artifactId>dropwizard-testing</artifactId>
-            <version>${dropwizard.version}</version>
+            <groupId>io.dropwizard.modules</groupId>
+            <artifactId>dropwizard-testing-junit4</artifactId>
+            <version>3.0.6</version>
```
Comment (alexbishop1) [inline pom.xml]: Should also be 3.0.7.

File: src/main/java/uk/gov/pay/connector/gateway/ClientFactory.java [hunk 7]
```diff
@@ -102,19 +101,15 @@ private HttpClientConnectionManager createConnectionManager(String gatewayName,
...
-                SystemDefaultDnsResolver.INSTANCE,
-                connectionTimeToLive.toMilliseconds(),
-                TimeUnit.MILLISECONDS,
-                format("%s.%s", gatewayName, operation)
-        );
+                        .build())
+                .connFactory(new ManagedHttpClientConnectionFactory())
+                .dnsResolver(SystemDefaultDnsResolver.INSTANCE)
+                .timeToLive(TimeValue.of(java.time.Duration.ofMillis(connectionTimeToLive.toMilliseconds())))
+                .name(format("%s.%s", gatewayName, operation)).build();
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/gateway/ClientFactory.java]: I assume this pattern comes is copied somewhere else where we’re already using HttpClient 5?

File: src/test/java/uk/gov/pay/connector/it/dao/TokenDaoJpaIT.java [hunk 8]
```diff
@@ -110,7 +108,7 @@ void deleteByCutOffDate_shouldDeleteOlderTokens() {
         final ZonedDateTime expiryThreshold = today.minusDays(7);
         tokenDao.deleteTokensOlderThanSpecifiedDate(expiryThreshold);
 
-        assertThat(tokenDao.findByTokenId("old-token"), isEmpty());
-        assertThat(tokenDao.findByTokenId("present-day-token"), isPresent());
+        assertThat(tokenDao.findByTokenId("old-token").isEmpty(), is(true));
+        assertThat(tokenDao.findByTokenId("present-day-token").isPresent(), is(true));
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/it/dao/TokenDaoJpaIT.java:112]: Good spot!

**Review by alexbishop1 — CHANGES_REQUESTED** [review]
Looking good! Just a few small tweaks and I think we’re there!

File: .github/dependabot.yml (hunk 1 above)
Comment (SandorArpa) [reply .github/dependabot.yml]: 👍

File: pom.xml (hunk 2 above)
Comment (SandorArpa) [reply pom.xml]: 👍

File: pom.xml (hunk 3 above)
Comment (SandorArpa) [reply pom.xml]: 👍

File: pom.xml (hunk 4 above)
Comment (SandorArpa) [reply pom.xml]: 👍

File: pom.xml (hunk 5 above)
Comment (SandorArpa) [reply pom.xml:383]: 👍

File: pom.xml (hunk 6 above)
Comment (SandorArpa) [reply pom.xml]: 👍

File: src/test/java/uk/gov/pay/connector/it/dao/TokenDaoJpaIT.java (hunk 8 above)
Comment (SandorArpa) [reply src/test/java/uk/gov/pay/connector/it/dao/TokenDaoJpaIT.java:112]: 👍

File: src/main/java/uk/gov/pay/connector/gateway/ClientFactory.java (hunk 7 above)
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/connector/gateway/ClientFactory.java]: No, I re-wrote it. We don't use this setup anywhere else.
Dropwizard 3:

`+- io.dropwizard:dropwizard-client:jar:3.0.7:compile`
`|  +- org.glassfish.jersey.core:jersey-client:jar:2.41:compile`
`|  +- org.apache.httpcomponents.client5:httpclient5:jar:5.3.1:compile`
`|  |  \- org.apache.httpcomponents.core5:httpcore5-h2:jar:5.2.4:compile`
`|  +- org.apache.httpcomponents.core5:httpcore5:jar:5.2.4:compile`
`|  \- io.dropwizard.metrics:metrics-httpclient5:jar:4.2.25:compile`

and it comes with `com.codahale.metrics.httpclient5.InstrumentedHttpClientConnectionManager` which depends on apache's client5.

File: src/main/java/uk/gov/pay/connector/gateway/ClientFactory.java [hunk 9]
```diff
@@ -102,19 +101,15 @@ private HttpClientConnectionManager createConnectionManager(String gatewayName,
             throw new RuntimeException("Unable to create SSL connection socket factory", e);
         }
 
-        return new InstrumentedHttpClientConnectionManager(
-                metricRegistry,
-                RegistryBuilder.<ConnectionSocketFactory>create()
+       return InstrumentedHttpClientConnectionManager.builder(metricRegistry)
+                .socketFactoryRegistry(org.apache.hc.core5.http.config.RegistryBuilder.<ConnectionSocketFactory>create()
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/gateway/ClientFactory.java]: I don’t think `RegistryBuilder` needs to be full qualified.

File: src/main/java/uk/gov/pay/connector/gateway/ClientFactory.java [hunk 10]
```diff
@@ -102,19 +101,15 @@ private HttpClientConnectionManager createConnectionManager(String gatewayName,
...
-                null,
-                SystemDefaultDnsResolver.INSTANCE,
-                connectionTimeToLive.toMilliseconds(),
-                TimeUnit.MILLISECONDS,
-                format("%s.%s", gatewayName, operation)
-        );
+                        .build())
+                .connFactory(new ManagedHttpClientConnectionFactory())
+                .dnsResolver(SystemDefaultDnsResolver.INSTANCE)
+                .timeToLive(TimeValue.of(java.time.Duration.ofMillis(connectionTimeToLive.toMilliseconds())))
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/gateway/ClientFactory.java]: We could avoid the confusion with the two different `Duration` types (argh!) by doing this:

```java
.timeToLive(TimeValue.ofMilliseconds(connectionTimeToLive.toMilliseconds()))
```

**Review by alexbishop1 — CHANGES_REQUESTED** [review]
Just a couple of final things and then we’re good!

File: src/main/java/uk/gov/pay/connector/gateway/ClientFactory.java (hunk 10 above)
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/connector/gateway/ClientFactory.java]: Good catch 👍

File: src/main/java/uk/gov/pay/connector/gateway/ClientFactory.java (hunk 9 above)
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/connector/gateway/ClientFactory.java]: 👍

**Review by alexbishop1 — DISMISSED** [review]
🥳

**Review by alexbishop1 — DISMISSED** [review]

File: pom.xml [hunk 11]
```diff
@@ -52,9 +180,13 @@
...
         <dependency>
             <groupId>uk.gov.service.payments</groupId>
-            <artifactId>logging</artifactId>
+            <artifactId>logging-dropwizard-3</artifactId>
             <version>${pay-java-commons.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.glassfish.jersey.connectors</groupId>
+            <artifactId>jersey-apache-connector</artifactId>
+        </dependency>
```
Comment (alexbishop1) [inline pom.xml]: This should be in the `<!-- Main dependencies that are imported from one of the BOMs…` bit.

File: pom.xml [hunk 12]
```diff
@@ -495,28 +454,6 @@
             <version>${jjwt.version}</version>
             <scope>runtime</scope>
         </dependency>
```
Comment (alexbishop1) [inline pom.xml]: All three of thre JJWT dependencies belong in the `<!-- Main dependencies that need explicit versions -->` bit.

**Review by alexbishop1 — CHANGES_REQUESTED** [review]
A couple of dependencies are in the wrong place in the POM (which makes no functional difference, just makes it a little harder to follow).

File: pom.xml (hunk 11 above)
Comment (SandorArpa) [reply pom.xml]: 👍

File: pom.xml (hunk 12 above)
Comment (SandorArpa) [reply pom.xml]: 🤦

**Review by alexbishop1 — APPROVED** [review]
🥳

**Pull request merged**